src/java.base/share/classes/com/sun/crypto/provider/BlockCipherParamsCore.java

```diff
@@ -108,7 +108,7 @@ void init(byte[] encoded, String decodingMethod) throws IOException {
 
     <T extends AlgorithmParameterSpec> T getParameterSpec(Class<T> paramSpec)
         throws InvalidParameterSpecException {
-        if (IvParameterSpec.class.isAssignableFrom(paramSpec)) {
+        if (paramSpec.isAssignableFrom(IvParameterSpec.class)) {
```

Comment:
The call to cast() is confusing.   But if the paramSpec is AlgorithmParameterSpec.class or Object.class, what's the expected behavior?  There are potential casting exception, I guess.  Maybe, a exactly class matching could be better.

Reply:
If so, then the `if` block will be true and the spec object is casted to your specified class (`AlgorithmParameterSpec.class` or `Object.class`) and it always succeeds.

This is exactly what I want to achieve. In fact, this bug and the other `getInstance(oid)` bug have the same root. I was trying to decode an algorithm identifier from its encoding. First, the encoding of the algorithm is in OID so `AlgorithmParameters.getInstance()` must support OID. Second, I want to get the spec from the parameters without knowing the algorithm name and the child `AlgorithmParametersSpec` class type, so `AlgorithmParameters::getParameterSpec` must support `AlgorithmParameterSpec.class` as the argument.

Otherwise, the program needs to know name and parameter spec type on all supported algorithms.

Reply:
The check ensures casting always succeeds. The fact that this has not been noticed for such a long time means everyone is using the exact subclass type when calling the method.

Reply:
Right, I suppose so.

Reply:
PBEKeyFactory.java, PBKDF2Core.java and PBKDF2HmacSHA1Factory.java also have isAssignableFrom() calls which seem backward. Perhaps covering them as well?

Reply:
> If so, then the `if` block will be true and the spec object is casted to your specified class (`AlgorithmParameterSpec.class` or `Object.class`) and it always succeeds.
> 
> This is exactly what I want to achieve.

Unfortunately, there is a returned value that we cannot return an object of any class.

Reply:
> > If so, then the `if` block will be true and the spec object is casted to your specified class (`AlgorithmParameterSpec.class` or `Object.class`) and it always succeeds.
> > This is exactly what I want to achieve.
> 
> Unfortunately, there is a returned value that we cannot return an object of any class.

Oh, I missed that the paramSpec should be of class T, which extends AlgorithmParameterSpec.  Then, I have no more concerns.